### PR TITLE
[AMORO-1845][Optimizer] Add calculate optimizer resource allocation on optimizer side

### LINKED ISSUE #1845: [Improvement]: Show correct resource occupation for external optimizer
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/arctic/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Now the resources reported by external optimizer are configured through the -m parameter. If you modify the resource without modifying the `-m` parameter, it will cause AMS to display wrong resource occupation.

### How should we improve?

Calculate resource utilization in optimizer, override `-m` parameter.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zhongqishang**
## Why are the changes needed?

fix #1845 

## Brief change log

  - *Add calculate optimizer resource allocation on optimizer side*

## How was this patch tested?
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Review by wangtaohz — COMMENTED** [review]
Thanks for your contribution! 👍
This PR also made me realize that the previous calculation of memory was wrong, in the case of TaskManager with slot>1, which is
```
    long memory = jobManagerMemory + taskManagerMemory * resource.getThreadCount();
```.

I think we should remove this erroneous code to avoid causing confusion for other developers

**Comment by wangtaohz** [comment]
Furthermore, `LocalOptimizer` may also adopt a similar approach to return memory usage during the optimizer's runtime, in order to unify the mechanism for calculating memory usage for optimizers.

How do you think about it? @zhongqishang

**Comment by zhongqishang** [comment]
> I think we should remove this erroneous code to avoid causing confusion for other developers

If this logic is removed, how should we set the value of `-m`?

> Furthermore, LocalOptimizer may also adopt a similar approach to return memory usage during the optimizer's runtime, in order to unify the mechanism for calculating memory usage for optimizers.

I agree. The user may submit the LocalOptimizer themselves and set an incorrect value.

**Comment by wangtaohz** [comment]
> If this logic is removed, how should we set the value of `-m`?

I think we can set it to 0 for `FlinkOptimizer`, and this `-m` config will be deprecated if `LocalOptimizer` no longer depends on it.

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/flink/FlinkOptimizer.java [hunk 1]
```diff
@@ -36,4 +44,25 @@ public static void main(String[] args) throws CmdLineException {
...
+    ReadableConfig configuration = env.getConfiguration();
+    MemorySize jobMemorySize = configuration.get(JobManagerOptions.TOTAL_PROCESS_MEMORY);
+    MemorySize taskMemorySize = configuration.get(TaskManagerOptions.TOTAL_PROCESS_MEMORY);
+    if (jobMemorySize == null || taskMemorySize == null) {
+      // Running locally will not get the configuration.
+      return;
+    }
+    int parallelism = config.getExecutionParallel();
+    int numberOfTaskSlots = configuration.get(TaskManagerOptions.NUM_TASK_SLOTS);
+    int memorySize = jobMemorySize.getMebiBytes() + (parallelism / numberOfTaskSlots) * taskMemorySize.getMebiBytes();
```
Comment (shidayang) [inline ams/optimizer/src/main/java/com/netease/arctic/optimizer/flink/FlinkOptimizer.java:58]: The TaskManager's count cannot be simply calculated using parallelism/numberOfTaskSlots. The capacity of the TaskManager is related to many configuration parameters in Flink. Can the calculation method used in Flink be reused here?

Comment (wangtaohz) [reply ams/optimizer/src/main/java/com/netease/arctic/optimizer/flink/FlinkOptimizer.java:58]: In the following code, when calculating the TaskManager's count, it takes into account rounding up the result of `parallelism/numberOfTaskSlots`.

Can you provide more details of mechanisms which may affect the count of TaskManagers? @shidayang

**Comment by zhongqishang** [comment]
> this -m config will be deprecated if LocalOptimizer no longer depends on it.

`localOptimize.sh` depends on the -m parameter setting `-Xmx`, I think it is still needed.

**Comment by wangtaohz** [comment]
> localOptimize.sh depends on the -m parameter setting -Xmx, I think it is still needed.

I checked the `localOptimize.sh`, 
```
JAVA_OPTS="-Xmx$2m -Dlog.home=${LOG_DIR} -Dlog.subdir=localOptimizer-$(date +%s)"
```
we could see that the `-Xmx` is set by the secondary parameter, which happens to be the memory after `-m`, but it has nothing to do with `-m`.

```
    return String.format("%s/bin/localOptimize.sh -m %s %s", getAMSHome(), memory,
        super.buildOptimizerStartupArgsString(resource));
```

So the `-m` can be deprecated, and after taht, we can directly use the `-Xmx` to set the memory. Of course, some modifications need to be made to `localOptimize.sh`.

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/flink/FlinkOptimizer.java (hunk 1 above)
Comment (shidayang) [reply ams/optimizer/src/main/java/com/netease/arctic/optimizer/flink/FlinkOptimizer.java:58]: The Flink code has been changed significantly and I couldn't find it，so this calculation method is also okay.

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/local/LocalOptimizer.java [hunk 2]
```diff
@@ -12,6 +12,11 @@ public class LocalOptimizer {
   public static void main(String[] args) throws CmdLineException {
     OptimizerConfig optimizerConfig = new OptimizerConfig(args);
     Optimizer optimizer = new Optimizer(optimizerConfig);
+
+    // calculate optimizer memory allocation
+    long memorySize = Runtime.getRuntime().maxMemory() / 1024 / 1024;
+    optimizerConfig.setMemorySize((int) memorySize);
+
```
Comment (wangtaohz) [inline ams/optimizer/src/main/java/com/netease/arctic/optimizer/local/LocalOptimizer.java:19]: I tested it locally.
image
Although there are some differences compared to -Xmx, I think the results are acceptable. 👍

Comment (zhongqishang) [reply ams/optimizer/src/main/java/com/netease/arctic/optimizer/local/LocalOptimizer.java:19]: Yes, my local test also has a little gap. And the gap is not big, it feels acceptable.

**Comment by wangtaohz** [comment]
We would appreciate it if you could provide some test results(like screenshots) to demonstrate the correctness of the logic for computing Flink optimizer memory, as it cannot be covered by unit tests. Could you help with this? @zhongqishang

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/LocalOptimizerContainer.java [hunk 3]
```diff
@@ -54,7 +53,7 @@ protected String buildOptimizerStartupArgsString(Resource resource) {
     long memoryPerThread = Long.parseLong(PropertyUtil.checkAndGetProperty(resource.getProperties(),
         JOB_MEMORY_PROPERTY));
     long memory = memoryPerThread * resource.getThreadCount();
-    return String.format("%s/bin/localOptimize.sh -m %s %s", getAMSHome(), memory,
+    return String.format("%s/bin/localOptimize.sh %s %s", getAMSHome(), memory,
```
Comment (wangtaohz) [inline ams/optimizer/src/main/java/com/netease/arctic/optimizer/LocalOptimizerContainer.java:56]: After remove `-m`, The unit test cases in `TestOptimizerConfig` also needs adjustment.

Comment (zhongqishang) [reply ams/optimizer/src/main/java/com/netease/arctic/optimizer/LocalOptimizerContainer.java:56]: Done.

**Comment by zhongqishang** [comment]
The external optimizer parameters are as follows:
```
-a thrift://172.16.58.151:1261 -p 5 -g flinkGroup -id abcbc38d-9774-46c7-8b31-8dda48d1cf0c -hb 60000 -eds -msz 2048 -dsp /tmp
```
Memory
image

Optimizer
image

**Review by wangtaohz — APPROVED** [review]
LGTM

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/flink/FlinkOptimizer.java [hunk 4]
```diff
@@ -36,4 +44,25 @@ public static void main(String[] args) throws CmdLineException {
       LOG.error("Execute flink optimizer failed", e);
     }
   }
+
+  private static void calcOptimizerMemory(OptimizerConfig config, StreamExecutionEnvironment env) {
+    ReadableConfig configuration = env.getConfiguration();
+    MemorySize jobMemorySize = configuration.get(JobManagerOptions.TOTAL_PROCESS_MEMORY);
+    MemorySize taskMemorySize = configuration.get(TaskManagerOptions.TOTAL_PROCESS_MEMORY);
+    if (jobMemorySize == null || taskMemorySize == null) {
+      // Running locally will not get the configuration.
```
Comment (zhoujinsong) [inline ams/optimizer/src/main/java/com/netease/arctic/optimizer/flink/FlinkOptimizer.java:53]: Is `Running locally` means deploying the flink job in the standalone cluster?

Besides, how can we calculate the memory usage in this situation if we have deprecated the `-m` configuration?

File: docs/admin-guides/managing-optimizers.md [hunk 5]
```diff
@@ -135,4 +135,4 @@ The description of the relevant parameters is shown in the following table:
 | -a AMS_THRIFT_SERVER_URL | The address of the AMS thrift service, for example: thrift://127.0.0.1:1261, can be obtained from the config.yaml configuration. |
 | -g OPTIMIZE_GROUP_NAME | Group name created in advance under external container. |
 | -p EXECUTOR_PARALLELISM | Optimizer parallelism usage. |
-| -m EXECUTOR_MEMORY | Optimizer memory usage. |
\ No newline at end of file
+| -m EXECUTOR_MEMORY | @Deprecated Optimizer memory usage. |
```
Comment (zhoujinsong) [inline docs/admin-guides/managing-optimizers.md:138]: We may remove this configuration from docs if we have decided to remove it in the future.

**Review by zhoujinsong — COMMENTED** [review]
@zhongqishang This PR seems good to me, but I still have some question, please take another look when you have time.

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/flink/FlinkOptimizer.java (hunk 4 above)
Comment (zhongqishang) [reply ams/optimizer/src/main/java/com/netease/arctic/optimizer/flink/FlinkOptimizer.java:53]: > Is `Running locally` means deploying the flink job in the standalone cluster?

`Running locally` means running the optimizer in the IDEA. This method is supported in all deployment modes.

> Besides, how can we calculate the memory usage in this situation if we have deprecated the `-m` configuration?

In this situation, there is indeed no coverage. We can also obtain the usage of jvm in the same way as the local optimizer. Do you think it is necessary?

Comment (zhoujinsong) [reply ams/optimizer/src/main/java/com/netease/arctic/optimizer/flink/FlinkOptimizer.java:53]: Ok, I don't think we will run the flink optimizer in the IDEA.

If we do, we still can use an argument to specify the memory usage.

**Review by zhoujinsong — APPROVED** [review]

**Pull request merged**